### PR TITLE
Add validation in listener to fix following bugs

### PR DESCRIPTION
**Pull request by Ram-srini**
1. TCFH-733 - Avalon is processing the null data and data hash input request in the indata section and giving incorrect results
2. TCFH-734 - WorkOrderGetResult fails with server error with the removal of WorkOrderid
3. TCFH-736 - responseTimeoutMSecs parameter in WorkOrderSubmit API is not working with special/ alphanumeric characters
4. TCFH-735 - WorkOrderSubmit fails with server error with the removal of requesterId
5. TCFH-703 - Direct model workOrderSubmit request is working without payload format in params field

Signed-off-by: Ramakrishna Srinivasamurthy <ramakrishna.srinivasamurthy@intel.com>

### CODE REVIEW

**Comment by manojgop** [comment]
Can you please mention what exactly is the issue fixed in commit message than only mentioning BUG IDs. these are internal jira and doesn't give enough context to a reviewer

File: listener/avalon_listener/tcs_listener.py [hunk 1]
```diff
@@ -104,6 +104,29 @@ def __init__(self, config):
...
+        for k in input_json.keys():
+            if k not in valid_keys:
+                logger.info("Invalid parameter in json rpc request")
+                return False
+        if "jsonrpc" in input_json:
+            if input_json["jsonrpc"] != "2.0":
+                logger.info("Invalid json rpc version")
+                return False
+        if "id" in input_json:
+            return type(input_json["id"]) == int
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply listener/avalon_listener/tcs_listener.py]: You mean to log it?

File: listener/avalon_listener/tcs_listener.py [hunk 2]
```diff
@@ -104,6 +104,29 @@ def __init__(self, config):
         for m in rpc_methods:
             self.dispatcher.add_method(m)
 
+    def _valid_json_rpc_request(self, input_json):
```
[the remark this answers is not in the document]
Comment (Ram-srini) [reply listener/avalon_listener/tcs_listener.py]: We are using jsonrpc package. It has JSONRPCResponseManager, but it is not throwing any exception for wrong json rpc methods.

File: listener/avalon_listener/tcs_listener.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pankajgoyal2) [reply listener/avalon_listener/tcs_listener.py]: yes

File: sdk/avalon_sdk/worker/worker_details.py [hunk 3]
```diff
@@ -205,9 +207,10 @@ def load_worker(self, worker_data):
             self.proof_data = json.loads(
                 worker_data['workerTypeData']['proofData'])
 
-        self.worker_id = crypto_utility.strip_begin_end_public_key(
+        w_id = crypto_utility.strip_begin_end_public_key(
             worker_data['workerTypeData']['verificationKey']) \
-            .encode("UTF-8").hex()
+            .encode("UTF-8")
+        self.worker_id = hashlib.sha256(w_id).hexdigest()
```
Comment (manojgop) [inline sdk/avalon_sdk/worker/worker_details.py]: Could you please add a comment what is this change for ? @rranjan3 had a similar done a change in a different PR.

File: listener/avalon_listener/tcs_work_order_handler.py [hunk 4]
```diff
@@ -99,42 +113,52 @@ def WorkOrderGetResult(self, **params):
...
+            if value:
+                response = json.loads(value)
+                if 'result' in response:
+                    return response['result']
+
+                # response without a result should have an error
+                err_code = response["error"]["code"]
+                err_msg = response["error"]["message"]
+                if err_msg == self.INVALID_WORKLOAD_MESSAGE:
+                    err_msg = 'Invalid workload id'
```
Comment (manojgop) [inline listener/avalon_listener/tcs_work_order_handler.py]: Better to check the err_code and then set the required err_msg.
```
 if err_code == xxx:
     err_msg = 'Invalid workload id'
```

File: listener/avalon_listener/tcs_work_order_handler.py [hunk 5]
```diff
@@ -55,7 +56,8 @@ def __init__(self, kv_helper, max_wo_count):
         self.workorder_count = 0
         self.max_workorder_count = max_wo_count
         self.workorder_list = []
-
+        self.INVALID_WORKLOAD_MESSAGE = \
+            'CreateWorkloadProcessor function returned null'
```
Comment (manojgop) [inline listener/avalon_listener/tcs_work_order_handler.py]: This doesn't look good. listener ideally shouldn't check for exact error message string of enclave manager. Instead use an error code.

File: listener/avalon_listener/tcs_listener.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply listener/avalon_listener/tcs_listener.py]: That's interesting.  In above link , the implementation pf `JSONRPCResponseManager.handle()`  checks for `JSONRPCInvalidRequestException` and `return JSONRPC20Response(error=JSONRPCInvalidRequest()._data)`

[the remark this answers is not in the document]
Comment (manojgop) [reply listener/avalon_listener/tcs_listener.py]: Looks like we are not checking for error in response object in function `_process_request()`. Check if response object has error attribute or only data attribute
```
response = JSONRPCResponseManager.handle(data, self.dispatcher)
return response.data
```

[the remark this answers is not in the document]
Comment (Ram-srini) [reply listener/avalon_listener/tcs_listener.py]: These exceptions should be thrown from handler functions, not outside. Otherwise it will return in http error 500.

File: listener/avalon_listener/tcs_work_order_handler.py (hunk 5 above)
Comment (Ram-srini) [reply listener/avalon_listener/tcs_work_order_handler.py]: This is what we get it from enclave manager. value error and message as CreateWorkloadProcessor function returned null

File: listener/avalon_listener/tcs_work_order_handler.py (hunk 4 above)
Comment (Ram-srini) [reply listener/avalon_listener/tcs_work_order_handler.py]: We won't get the error code from enclave manager for this error.

File: listener/avalon_listener/tcs_listener.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply listener/avalon_listener/tcs_listener.py]: Yes.  Need to check if response object returned by `JSONRPCResponseManager.handle()` has error attribute if JSON RPC request is invalid and then we should handle the scenario for error. It's better to use library provided error handling for malformed JSON RPC request  than writing a custom function

File: listener/avalon_listener/tcs_work_order_handler.py (hunk 5 above)
Comment (manojgop) [reply listener/avalon_listener/tcs_work_order_handler.py]: Listener need to check only for error code enclave manager. It should not compare the error message from enclave manager. If you need specific error code then we need to modify enclave manager to return the specific error code

File: listener/avalon_listener/tcs_work_order_handler.py (hunk 4 above)
Comment (manojgop) [reply listener/avalon_listener/tcs_work_order_handler.py]: Thought so. @manju956 Can we return a new error code from enclave manager for this scenario. Please suggest where to where to make this change.

Comment (manju956) [reply listener/avalon_listener/tcs_work_order_handler.py]: Enclave manager don't add any error code as such. New error codes can be added in tcf_error.h and we need to define a class for the new error code in error.h.
Also, we need to update the enclave code to return new error code when invalid workload_id has been identified by the enclave.

Comment (manojgop) [reply listener/avalon_listener/tcs_work_order_handler.py]: Yeah. We need to do define new error codes corresponding to workload id and enclave manager should return that. This should be checked by listener

File: listener/avalon_listener/tcs_listener.py (hunk 1 above)
Comment (manojgop) [inline listener/avalon_listener/tcs_listener.py]: Please check https://www.jsonrpc.org/specification. An identifier established by the Client that MUST contain a String, Number, or NULL value if included. 

There are other error conditions for json rpc request which are not validated by this function. I would suggest to use library provided error handling than writing a custom error handling function for validating json rpc request. 
`response = JSONRPCResponseManager.handle()` returns a response object which will contain an error field in case of error. For invalid json rpc request,  `response.error` will look like this -> `{'code': -32600, 'message': 'Invalid Request'}`

Comment (Ram-srini) [reply listener/avalon_listener/tcs_listener.py]: In our spec everywhere id is mentioned as integer. Then we should update our spec as well.

File: sdk/avalon_sdk/worker/worker_details.py (hunk 3 above)
Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: Done

File: listener/avalon_listener/tcs_listener.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ram-srini) [reply listener/avalon_listener/tcs_listener.py]: not required

File: sdk/avalon_sdk/worker/worker_details.py (hunk 3 above)
Comment (manojgop) [reply sdk/avalon_sdk/worker/worker_details.py]: The worker_id stored in block chain as part of worker details should already be in 32 bytes as enclave manager is storing it in 32 bytes in shared KV DB -> https://github.com/hyperledger/avalon/blob/master/enclave_manager/avalon_enclave_manager/enclave_manager.py#L99. So why we need to duplicate the calculation of worker id again ?

Comment (manojgop) [reply sdk/avalon_sdk/worker/worker_details.py]: load_worker() function should take worker_id as argument instead of recalculating it here

Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: why do we need to pass worker_id to load_worker again? worker_data['workerTypeData']['verificationKey'] which is enclave public key, that we are converting it to sha256. Don't you think it is also duplicate? First of all why worker_id calculation is there in load_worker() earlier as well? Earlier also it was duplication calculation.

Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: Neater approach would be to use the worker_id from the worker_lookup() API rather than from the worker_details?

Comment (manojgop) [reply sdk/avalon_sdk/worker/worker_details.py]: Verification Key is the public part of signing key which is different from worker id.  So verification key is not duplicate. Earlier logic was also not correct. Can we fix that along with this bug fix PR ? Else please add a TODO comment in code and raise an issue which can be fixed in different PR.

Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: in load_worker(), we are calculating worker_id from ['workerTypeData']['verificationKey']. Why it was not taken care earlier itself? How come it is fine now if that logic is incorrect?

Comment (manojgop) [reply sdk/avalon_sdk/worker/worker_details.py]: Pass worker_id in constructor of workerDetails

Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: Created the issue to remote the worker id calculation in the sdk. https://github.com/hyperledger/avalon/issues/359

Comment (Ram-srini) [reply sdk/avalon_sdk/worker/worker_details.py]: > Pass worker_id in constructor of workerDetails

No need to worker_id in workerDetails. User should use the worker id got from worker_lookup().

Comment (manojgop) [reply sdk/avalon_sdk/worker/worker_details.py]: I mean when we crate WorkerDetails object, we should know for which worker_id this object is being created for. For that we can use worker_id we got from worker lookup.
